ci: Split Experimental tests from Universe tests

piraces opened the pull request:
Closes #2375 

A new workflow has been created to run the experimental packages "x" inside "universe.dagger.io".
This workflow is named `test-experimental-universe.yml` and its based on `test-universe.yml`.

The `package.json` in the "universe.dagger.io" folder has been modified to add another script to only test the experimental packages and the original "test" script now ignores the tests in folder "x".

Main `Makefile` has been modified to add a new task for this.

After reviewing and merging this PR when all is ready, we need to ensure that in the repo settings, in "Branch protection rules", the "Status checks that are required" does not contain this workflow, so PRs can be merged when this tests fail:

![Settings to modify in repo](https://user-images.githubusercontent.com/11027628/168479258-179dc5de-36ad-4182-9793-4f5f24e94bd4.png)


Signed-off-by: piraces <raul.piraces@gmail.com>

TomChv commented on .github/workflows/test-experimental-universe.yml:
```diff
@@ -0,0 +1,76 @@
+      - "go.mod"
+      - "go.sum"
+      - ".github/workflows/test-experimental-universe.yml"
+      - "!docs/**"
```
I think we should trigger that ci on change in [`docs/tests/**`](https://github.com/dagger/dagger/tree/main/docs/tests) even if they are not running for now.

piraces replied:
Agree!

TomChv commented on .github/workflows/test-experimental-universe.yml:
```diff
@@ -0,0 +1,76 @@
+      - "go.mod"
+      - "go.sum"
+      - ".github/workflows/test-experimental-universe.yml"
+      - "!docs/**"
```
I think we should trigger that ci on change in [`docs/tests/**`](https://github.com/dagger/dagger/tree/main/docs/tests) even if they are not running for now.

TomChv commented on .github/workflows/test-experimental-universe.yml:
```diff
@@ -0,0 +1,76 @@
+  DAGGER_CACHE_BASE: dagger-ci-experimental-universe
+
+jobs:
+  experimental-universe-europa-tests:
```
just `experimental-universe-tests` ;)

TomChv commented on Makefile:
```diff
@@ -69,6 +69,11 @@ universe-test: dagger-debug # Run universe tests
 	yarn --cwd "./pkg/universe.dagger.io" install
 	DAGGER_BINARY="$(shell pwd)/cmd/dagger/dagger-debug" yarn --cwd "./pkg/universe.dagger.io" test
 
+.PHONY: experimental-universe-test
```
For consistency: `experimental-universe-tests`

TomChv reviewed (commented):
Thanks for your contributions! I left minors comments to improve this PR :rocket:

piraces commented:
@TomChv Thanks for the tips! Just committed the changes 👍

samalba commented:
Given the PR #2442 (basically moving everything to a dagger plan), do we still need this?

piraces commented:
@samalba From my point of view I think it can be useful in order to not block pull requests that have failed tests from the experimental packages.
Maybe we need to "sync" this behaviour in the new dagger CI plan, having a separate section for the experimental packages which the new workflow will execute.

piraces commented:
After the PR #2465, which allows to filter universe tests, I have adapted this PR taking into account the #2442 plan. So this will be need to be ported to the dagger plan as soon as #2442 is complete.

It will make the project to acomplish the original issue and be synced with the current state of the CI.

What do you think @samalba @TomChv ?

sipsma commented on pkg/universe.dagger.io/package.json:
```diff
@@ -1,7 +1,8 @@
   "scripts": {
-    "test": "bats --report-formatter junit --print-output-on-failure --jobs 4 $(find ${TESTDIR:-.} -type f -name '*.bats' -not -path '*/node_modules/*')"
+    "test": "bats --report-formatter junit --print-output-on-failure --jobs 4 $(find ${TESTDIR:-.} -type f -name '*.bats' -not -path '*/node_modules/*' -not -path '*/x/*')",
+    "test-experimental": "bats --report-formatter junit --print-output-on-failure --jobs 4 $(find ${TESTDIR:-.} -type f -name '*.bats' -not -path '*/node_modules/*')"
```
```suggestion
    "test-experimental": "bats --report-formatter junit --print-output-on-failure --jobs 4 $(find ${TESTDIR:-./x} -type f -name '*.bats' -not -path '*/node_modules/*')"
```

We should only run the experimental tests as part of this step (right now if you look in CI it includes all of them in this step)

sipsma commented on Makefile:
```diff
@@ -69,6 +69,11 @@ universe-test: dagger-debug # Run universe tests
+.PHONY: experimental-universe-tests
+experimental-universe-tests: dagger-debug # Run experimental universe tests
+	yarn --cwd "./pkg/universe.dagger.io" install
+	TESTDIR=$(UNIVERSE_EXPERIMENTAL_TESTDIR) DAGGER_BINARY="$(shell pwd)/cmd/dagger/dagger-debug" yarn --cwd "./pkg/universe.dagger.io" test-experimental
```
Unless I'm missing something, `UNIVERSE_EXPERIMENTAL_TESTDIR` doesn't exist. If we make the change mentioned in my other comment, it shouldn't need to either so I think we can just remove this.

piraces replied:
Agree. I removed it.

sipsma reviewed (commented):
Looks good, thanks for the contribution! Two quick comments

sipsma commented:
The timeout you are getting in the universe tests should be fixed by this: https://github.com/dagger/dagger/pull/2495

piraces commented:
Rebasing... thanks!

samalba commented:
@piraces I was about to merge and noticed the commits are weird. Maybe the rebase was not done correctly?

If you go to `main`, do a `git pull`, then switch to this branch and `git rebase main && git push -f` it should solve this. At the end, you should only see your own commits in this PR. Let me know if you need help.

sipsma reviewed (approved):
LGTM, thanks again!

piraces commented:
Hi @samalba I have squashed all my commits into a single one and rebased from main. I think now its correct.

I have also made some changes in order to "adapt" to the latest CI commit and run the experimental tests with dagger as the universe ones.

I hope this is now correct and ok regarding latest changes 😃
